libnavui-util/src/main/java/com/mapbox/navigation/ui/utils/internal/lifecycle/ViewLifecycleRegistry.kt

```diff
@@ -0,0 +1,89 @@
+
+/**
+ * A [LifecycleRegistry] that merges the [hostingLifecycleOwner]'s [Lifecycle] (of an Activity or Fragment)
+ * with the [View]'s drawing surface lifecycle.
+ *
+ * This lifecycle is useful for continuously running tasks that obtain the data and draw it in the [View],
+ * for example by collecting [Flow]s and drawing results in the nested views or on the map.
+ *
+ * In general, this lifecycle reflects the [hostingLifecycleOwner]'s [Lifecycle] whenever the `View` is attached
+ * but whenever the `View` is detached and we want to stop the coroutines tied to the lifecycle's scope,
+ * this lifecycle will transition to [Lifecycle.State.CREATED] (and invoke [Lifecycle.Event.ON_STOP])
+ * even if the hosting Activity or Fragment is still running.
+ *
+ * This lifecycle reaches [Lifecycle.State.DESTROYED] only when the [hostingLifecycleOwner] is destroyed.
+ *
+ * @see keepExecutingWhenStarted
+ */
+@UiThread
+class ViewLifecycleRegistry(
+    private val hostingLifecycleOwner: LifecycleOwner,
+    localLifecycleOwner: LifecycleOwner
+) : LifecycleRegistry(localLifecycleOwner) {
+
+    private var isAttached = false
+
+    private val hostingLifecycleObserver = LifecycleEventObserver { _, event ->
+        val isAtLeastCreated = currentState.isAtLeast(State.CREATED)
+        if (isAttached || (isAtLeastCreated && event == Event.ON_DESTROY)) {
+            handleLifecycleEvent(event)
+        }
+    }
+
+    init {
+        hostingLifecycleOwner.lifecycle.addObserver(hostingLifecycleObserver)
+    }
+
+    /**
+     * Call whenever [View.onAttachedToWindow] of the target [View] is called.
+     */
+    fun onAttachedToWindow() {
+        check(!isAttached) {
+            "onAttachedToWindow called without onDetachedFromWindow being called first"
+        }
+        currentState = hostingLifecycleOwner.lifecycle.currentState
+        isAttached = true
+    }
+
+    /**
+     * Call whenever [View.onDetachedFromWindow] of the target [View] is called.
+     */
+    fun onDetachedFromWindow() {
+        check(isAttached) {
+            "onDetachedFromWindow called without onAttachedToWindow being called first"
+        }
+        isAttached = false
+        if (hostingLifecycleOwner.lifecycle.currentState.isAtLeast(State.STARTED)) {
+            currentState = State.CREATED
+        }
+    }
+}
+
+/**
+ * Launches a coroutine whenever this lifecycle reaches [Lifecycle.State.STARTED] or higher.
+ * If the lifecycle state triggers [Lifecycle.Event.ON_STOP] or beyond, the launched coroutines are canceled.
+ *
+ * If the state reaches [Lifecycle.State.STARTED] again, the coroutines are restarted.
+ *
+ * This is useful when paired with [ViewLifecycleRegistry] because the coroutines will keep being canceled and restarted
+ * as the target [View] is detached and re-attached to window.
+ *
+ * This function uses [repeatOnLifecycle] under the hood.
+ */
+fun LifecycleOwner.keepExecutingWhenStarted(block: suspend CoroutineScope.() -> Unit) =
```

Comment:
another place for this could be `UIComponent`, that way it's scoped to only UIComponent implementations

Reply:
I figured it might be used in other places as well, not tied to drop-in's UI components.